### ISSUE TITLE
[Bug] train with the MlflowLoggerHook will have exception of "Invalid metric name"

### ISSUE THREAD

**Issue by RobinChiu** [open-mmlab/mmocr#1558]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmocr/issues) and [Discussions](https://github.com/open-mmlab/mmocr/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (0.x)](https://github.com/open-mmlab/mmocr) or [latest version (1.x)](https://github.com/open-mmlab/mmocr/tree/dev-1.x).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

main branch https://github.com/open-mmlab/mmocr

### Environment

/opt/conda/lib/python3.7/site-packages/mmcv/__init__.py:21: UserWarning: On January 1, 2023, MMCV will release v2.0.0, in which it will remove components related to the training process and add a data transformation module. In addition, it will rename the package names mmcv to mmcv-lite and mmcv-full to mmcv. See https://github.com/open-mmlab/mmcv/blob/master/docs/en/compatibility.md for more details.
  'On January 1, 2023, MMCV will release v2.0.0, in which it will remove '
sys.platform: linux
Python: 3.7.13 (default, Mar 29 2022, 02:18:16) [GCC 7.5.0]
CUDA available: True
GPU 0: NVIDIA GeForce GTX 1080 Ti
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.12.1
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.3.2  (built against CUDA 11.5)
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.3.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.12.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.13.1
OpenCV: 4.6.0
MMCV: 1.7.0
MMCV Compiler: GCC 9.3
MMCV CUDA Compiler: 11.3
MMOCR: 0.6.3+26bc471

### Reproduces the problem - code sample

root@1511067ecba8:/app/mmocr# git diff configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py
diff --git a/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py b/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py
index 467903f..5e41888 100644
--- a/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py
+++ b/configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py
@@ -13,7 +13,7 @@ train_pipeline_r18 = {{_base_.train_pipeline_r18}}
 test_pipeline_1333_736 = {{_base_.test_pipeline_1333_736}}
 
 data = dict(
-    samples_per_gpu=16,
+    samples_per_gpu=4,
     workers_per_gpu=8,
     val_dataloader=dict(samples_per_gpu=1),
     test_dataloader=dict(samples_per_gpu=1),
@@ -31,3 +31,10 @@ data = dict(
         pipeline=test_pipeline_1333_736))
 
 evaluation = dict(interval=100, metric='hmean-iou')
+
+log_config = dict(
+    interval=5,
+    hooks=[
+        dict(type='TextLoggerHook'), 
+        dict(type='MlflowLoggerHook', exp_name='icdar2015'), 
+    ])
root@1511067ecba8:/app/mmocr# 

### Reproduces the problem - command or script

root@1511067ecba8:/app/mmocr# python tools/train.py configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py --work-dir dbnet

### Reproduces the problem - error message

2022-11-18 08:53:47,946 - mmocr - INFO - Epoch(val) [100][500]  0_hmean-iou:recall: 0.0048, 0_hmean-iou:precision: 0.0002, 0_hmean-iou:hmean: 0.0004
Traceback (most recent call last):
  File "tools/train.py", line 243, in <module>
    run_train_cmd(args)
  File "tools/train.py", line 238, in run_train_cmd
    meta=meta)
  File "/app/mmocr/mmocr/apis/train.py", line 155, in train_detector
    runner.run(data_loaders, cfg.workflow)
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 136, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 58, in train
    self.call_hook('after_train_epoch')
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 317, in call_hook
    getattr(hook, fn_name)(self)
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/hooks/logger/base.py", line 164, in after_train_epoch
    self.log(runner)
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/dist_utils.py", line 144, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/mmcv/runner/hooks/logger/mlflow.py", line 79, in log
    self.mlflow.log_metrics(tags, step=self.get_iter(runner))
  File "/opt/conda/lib/python3.7/site-packages/mlflow/tracking/fluent.py", line 675, in log_metrics
    MlflowClient().log_batch(run_id=run_id, metrics=metrics_arr, params=[], tags=[])
  File "/opt/conda/lib/python3.7/site-packages/mlflow/tracking/client.py", line 1099, in log_batch
    self._tracking_client.log_batch(run_id, metrics, params, tags)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/tracking/_tracking_service/client.py", line 415, in log_batch
    self.store.log_batch(run_id=run_id, metrics=metrics_batch, params=[], tags=[])
  File "/opt/conda/lib/python3.7/site-packages/mlflow/store/tracking/file_store.py", line 1002, in log_batch
    _validate_batch_log_data(metrics, params, tags)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/utils/validation.py", line 317, in _validate_batch_log_data
    _validate_metric(metric.key, metric.value, metric.timestamp, metric.step)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/utils/validation.py", line 120, in _validate_metric
    _validate_metric_name(key)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/utils/validation.py", line 97, in _validate_metric_name
    INVALID_PARAMETER_VALUE,
mlflow.exceptions.MlflowException: Invalid metric name: 'val/0_hmean-iou:recall'. Names may only contain alphanumerics, underscores (_), dashes (-), periods (.), spaces ( ), and slashes (/).

### Additional information

train should finish and get the model and mlflow related information.

**Comment by Harold-lkk** [comment]
Thanks for your report! 
The bug is because MIFlow does not support the name containing the colon(:).In https://github.com/open-mmlab/mmocr/blob/d052f43450e39a05807038224544d42bb9f8bbbb/mmocr/core/evaluation/hmean.py#L169-171, the colon will be auto added into the metric name. 
You can replace ':' with '-' or '_', like
```python
        eval_results[metric + ':recall'] = best_result['recall']
        eval_results[metric + ':precision'] = best_result['precision']
        eval_results[metric + ':hmean'] = best_result['hmean']
```
to 
```python
        eval_results[metric + '_recall'] = best_result['recall']
        eval_results[metric + '_precision'] = best_result['precision']
        eval_results[metric + '_hmean'] = best_result['hmean']

**Comment by RobinChiu** [comment]
Thanks for the answer. I am trying the temporarily solution. 
But I think the metric should not use the ":" in the name. 
It's not easy to read or parse even in the log file. 

ex. 
In the log the ":" should split "metric" and "value". Once using the ":" in the metric name, it will be confuse. 

2022-11-15 04:16:32,051 - mmocr - INFO - Epoch(val) [1200][500]     0_hmean-iou:recall: 0.7636, 0_hmean-iou:precision: 0.8748, 0_hmean-iou:hmean: 0.8154

**Comment by gaotongxiao** [comment]
That's right, so we have fixed this issue in our [1.0](https://github.com/open-mmlab/mmocr/tree/dev-1.x) version. The logs in MMOCR 1.0 now look like: 

```bash
2022/08/25 22:30:00 - mmengine - INFO - Epoch(val) [20][500/500]  icdar/precision: 0.0000  icdar/recall: 0.0014  icdar/hmean: 0.0001
```

BTW, the 1.x branch will become the main branch very soon, and you are welcome to check it out. (But `MlflowLoggerHook` has been implemented in the latest architecture yet).
